### PR TITLE
Add RESTful API sample with DataMapper

### LINKED ISSUE #4600: Create a RESTful API sample with Data mapper
**Description:**

> $Subject

**Describe your task(s)**

- Create a sample ballerina application
- Add the samples to `ballerina-guides/intergration-samples`

### PR DESCRIPTION
**Pull request by TharmiganK**
> Fixes https://github.com/ballerina-platform/ballerina-standard-library/issues/4600

### CODE REVIEW

File: vs-code-home-page-sample/main.bal [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    }[] courses;
+    int totalCredits;
+    string visaType;
+};
+
+const D_TIER_4_VISA = "D tier-4";
+
+var totalCredits = function(int total, record {string id; string name; int credits;} course) returns int => total + (course.id.startsWith("CS") ? course.credits : 0);
+
+function transform(Person person, Course[] courses) returns Student => let var isForeign = person.country != "LK" in {
```
Comment (shafreenAnfar) [inline vs-code-home-page-sample/main.bal]: Lets use this in the service.

Comment (TharmiganK) [reply vs-code-home-page-sample/main.bal]: Ok 👍🏼

Comment (TharmiganK) [reply vs-code-home-page-sample/main.bal]: Fixed via [20e1fb0](https://github.com/ballerina-guides/integration-samples/pull/27/commits/20e1fb0de69931807521ff4a7a8344bdf46a5a14)

File: vs-code-home-page-sample/Ballerina.toml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+[package]
+org = "tharmigan"
```
Comment (anupama-pathirage) [inline vs-code-home-page-sample/Ballerina.toml]: need to change the org name.

File: vs-code-home-page-sample/Ballerina.toml [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+name = "vs_code_home_page_sample"
```
Comment (anupama-pathirage) [inline vs-code-home-page-sample/Ballerina.toml]: Add proper name for this usecase, don't say vscode sample

**Comment by praneesha** [comment]
@TharmiganK - We need to change the name of the sample to a meaningful one.

File: vs-code-home-page-sample/Ballerina.toml (hunk 3 above)
Comment (TharmiganK) [reply vs-code-home-page-sample/Ballerina.toml]: Fixed via https://github.com/ballerina-guides/integration-samples/pull/27/commits/e77c1725524b76903cfb174ff40803d7df44531b

File: vs-code-home-page-sample/Ballerina.toml (hunk 2 above)
Comment (TharmiganK) [reply vs-code-home-page-sample/Ballerina.toml]: Fixed via https://github.com/ballerina-guides/integration-samples/pull/27/commits/e77c1725524b76903cfb174ff40803d7df44531b

File: restful_api_with_data_mapper/main.bal [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        string title;
+        int credits;
+    }[] courses;
+    int totalCredits;
+    string visaType;
+};
+
+const D_TIER_4_VISA = "D tier-4";
+
+var totalCredits = function(int total, record {string id; string name; int credits;} course) returns int => total + (course.id.startsWith("CS") ? course.credits : 0);
```
Comment (anupama-pathirage) [inline restful_api_with_data_mapper/main.bal]: Break the line, usually 120 char limit.

File: restful_api_with_data_mapper/main.bal [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+type Course record {
+    string id;
+    string name;
+    int credits;
+};
+
+type Student record {
+    string id;
+    string fullName;
+    string age;
```
Comment (anupama-pathirage) [inline restful_api_with_data_mapper/main.bal]: Is it meaningful to use age as string?

File: restful_api_with_data_mapper/main.bal [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        country: "LK"
+    },
+    {
+        id: "1002",
+        firstName: "Jane",
+        lastName: "Doe",
+        age: 23,
+        country: "US"
+    }
+];
```
Comment (anupama-pathirage) [inline restful_api_with_data_mapper/main.bal]: Can we have these definitions at the top (before the transform function) for clarity?

File: restful_api_with_data_mapper/main.bal [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    }[] courses;
+    int totalCredits;
+    string visaType;
+};
+
+const D_TIER_4_VISA = "D tier-4";
+
+var totalCredits = function(int total, record {string id; string name; int credits;} course) returns int => total + (course.id.startsWith("CS") ? course.credits : 0);
+
+function transform(Person person, Course[] courses) returns Student => let var isForeign = person.country != "LK" in {
```
Comment (anupama-pathirage) [inline restful_api_with_data_mapper/main.bal]: Shall we use meaningful name than `transform` as the function name? Even datamapper UI allows to define custome name IINM.

**Review by anupama-pathirage — CHANGES_REQUESTED** [review]

**Comment by praneesha** [comment]
@TharmiganK - Shall we use a dash instead of an underscore when naming the project (i.e., `restful-api-with-data-mapper`) to be consistent with the other samples in https://github.com/ballerina-guides?

**Comment by praneesha** [comment]
@TharmiganK - Can we add the GSheet connector also to the sample as done in the `Consuming services` sample in https://ballerina.io/?

This is required to generate a comprehensive sequence diagram.

File: restful_api_with_data_mapper/main.bal (hunk 4 above)
Comment (TharmiganK) [reply restful_api_with_data_mapper/main.bal]: Fixed via [74db4ed](https://github.com/ballerina-guides/integration-samples/pull/27/commits/74db4ed66a889c18321790d55a4092e0647cf9b3)

File: restful_api_with_data_mapper/main.bal (hunk 6 above)
Comment (TharmiganK) [reply restful_api_with_data_mapper/main.bal]: Fixed via [74db4ed](https://github.com/ballerina-guides/integration-samples/pull/27/commits/74db4ed66a889c18321790d55a4092e0647cf9b3)

File: restful_api_with_data_mapper/main.bal (hunk 5 above)
Comment (TharmiganK) [reply restful_api_with_data_mapper/main.bal]: Fixed via [74db4ed](https://github.com/ballerina-guides/integration-samples/pull/27/commits/74db4ed66a889c18321790d55a4092e0647cf9b3)

File: restful_api_with_data_mapper/main.bal (hunk 7 above)
Comment (TharmiganK) [reply restful_api_with_data_mapper/main.bal]: Fixed via [d947cac](https://github.com/ballerina-guides/integration-samples/pull/27/commits/d947cacd24449500d642f90a284c9e224c1d244c)

File: restful-api-with-data-mapper/Config.toml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+sheetsAccessToken = "<Google Sheets API access token>"
+spreadSheetId = "<Google spreadsheet ID>"
+
+# Optional
+# sheetName = "student-enrollments"
```
Comment (anupama-pathirage) [inline restful-api-with-data-mapper/Config.toml]: For other samples we didn't commit the config toml. As a best practice it was added to .gitignore file. And README is updated with sample config.toml

Comment (TharmiganK) [reply restful-api-with-data-mapper/Config.toml]: Sure, will remove this and add a README

Comment (TharmiganK) [reply restful-api-with-data-mapper/Config.toml]: Fixed via [0961891](https://github.com/ballerina-guides/integration-samples/pull/27/commits/09618912461824f0bdd946319b01c6677c5f36d5)

**Review by anupama-pathirage — APPROVED** [review]

**Pull request merged**